### PR TITLE
Replace plugins by webhook

### PR DESCRIPTION
**Pull request by jasiu001**
Related to [#2791](https://github.com/kyma-project/kyma/issues/2791)

### CODE REVIEW

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	for _, v := range h.UpdateValidators {
+		admission.InjectDecoderInto(d, v)
+	}
+
+	return nil
+}
+
+// InjectClient injects the client into the handlers
+func (h *AdmissionHandler) InjectClient(c client.Client) error {
+	h.client = c
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]: not used, can be removed

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	return &AdmissionHandler{
+		CreateValidators: []Validator{&AccessToBroker{}},
+		UpdateValidators: []Validator{&AccessToBroker{}},
+	}
+}
+
+// Handle handles admission requests.
+func (h *AdmissionHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	traced := webhookutil.NewTracedLogger(req.UID)
+	traced.Infof("Start handling AdmissionHandler operation: %s for %s: %q", req.Operation, req.Kind.Kind, req.Name)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]:
```suggestion
	traced.Infof("Start handling validation operation: %s for %s: %q", req.Operation, req.Kind.Kind, req.Name)
```

`"Start handling AdmissionHandler operation` -> does not say if it's a validation or mutation, also `AdmissionHandler` does not bring additional value 


logs
```
I0402 14:25:19.538371       1 traced_logger.go:42] [ReqUID: 248cf598-5553-11e9-b571-fa82222f98b1 ]: Start handling operation: UPDATE for ServiceInstance: "si"
I0402 14:25:19.600992       1 traced_logger.go:42] [ReqUID: 248cf598-5553-11e9-b571-fa82222f98b1 ]: Completed successfully operation: UPDATE for ServiceInstance: "si"
I0402 14:25:19.602659       1 traced_logger.go:42] [ReqUID: 2496cab3-5553-11e9-b571-fa82222f98b1 ]: Start handling AdmissionHandler operation: UPDATE for ServiceInstance: "si"
I0402 14:25:19.602815       1 traced_logger.go:52] [ReqUID: 2496cab3-5553-11e9-b571-fa82222f98b1 ]: Starting validation - DenyPlanChangeIfNotUpdatable
I0402 14:25:19.602886       1 traced_logger.go:42] [ReqUID: 2496cab3-5553-11e9-b571-fa82222f98b1 ]: Completed successfully AdmissionHandler operation: UPDATE for ServiceInstance: "si"
```

as you can see it's not possible to determine which log "Start (..)" belongs to which logic.

Please also fix other places (`traced.Infof("Completed successfully ...` also needs to be adjusted)

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// AccessToBroker handles ClusterServiceBroker validation
+type AccessToBroker struct {
+	decoder *admission.Decoder
+	client  client.Client
+}
+
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go:38]: please add
```go
var _ admission.DecoderInjector = &AccessToBroker{}
var _ inject.Client = &AccessToBroker{}
```

same suggestion for other validators

## Background

Why it's so important? 
When you will not implement the write interface then the client will not be injected and you will get a panic in code when executing ` h.client.Create(ctx, sar)`

Using  `var _ inject.Client = &AccessToBroker{}` ensures that you implemented correct interface. What's more when you will update `controller-runtime` and they will change the interface then the compilation will fail

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+			Groups: user.Groups,
+			Extra:  convertToSARExtra(user.Extra),
+			UID:    user.UID,
+		},
+	}
+
+	err := h.client.Create(ctx, sar)
+	if err != nil {
+		traced.Errorf("Could not create SubjectAccessReview for %s %q: %v", csb.Kind, csb.Name, err)
+		return nil
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go]: you should return error here. You cannot allow processing broker which wasn't validated

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+			response := handler.Handle(context.TODO(), request)
+
+			// Then
+			assert.True(t, response.AdmissionResponse.Allowed)
+		})
+	}
+}
+
+func TestAdmissionHandlerAccessToBrokerDenied(t *testing.T) {
+	//Given
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go]:
```suggestion
	// given
```

I know it's nazi, but let's have a common pattern. In other pkg we using lowercase and space after `//`. Please unify that

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	decoder, err := admission.NewDecoder(scheme.Scheme)
+	require.NoError(t, err)
+
+	tests := map[string]struct {
+		operation admissionv1beta1.Operation
+		object    []byte
+	}{
+		"Request for Create ClusterServiceBroker should be allowed": {
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go]:
```suggestion
		"Request for Create ClusterServiceBroker without AuthInfo should be allowed": {
```
right now you missing why it should be allowed, same suggestion for the case below

File: pkg/webhook/servicecatalog/servicebinding/validation/handle_reference.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+// fail ServiceBinding operation if the ServiceInstance is marked for deletion
+// This feature was copied from Service Catalog admission plugin https://github.com/kubernetes-incubator/service-catalog/blob/v0.1.41/plugin/pkg/admission/servicebindings/lifecycle/admission.go
+// If you want to track previous changes please check there.
+func (h *ReferenceDeletion) Validate(ctx context.Context, req admission.Request, sb *sc.ServiceBinding, traced *webhookutil.TracedLogger) error {
+	instanceRef := sb.Spec.InstanceRef
+	instance := &sc.ServiceInstance{}
+
+	err := h.client.Get(ctx, types.NamespacedName{Namespace: sb.Namespace, Name: instanceRef.Name}, instance)
+	if err != nil {
+		traced.Errorf("Could not get ServiceInstance by name %q: %v", instanceRef.Name, err)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handle_reference.go]: previously when instance cannot be fetched then error wasn't returned. Please remember that in this task we are porting functionality and the behavior need to be the same as it was previously. Right now we do not want to add any new functionality or do any bug fixing. 

What's more such action could be on purpose, e.g. right now I cannot create an SB for not existing ServiceInstance. In the previous version I was able to do that -> Create ServiceBinding then create Service Instance and after that controller successful processed both kinds (so some small race condition for creating those objs were allowed).

File: pkg/webhook/servicecatalog/servicebinding/validation/handle_reference.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		return err
+	}
+
+	if instance.DeletionTimestamp != nil {
+		traced.Infof(
+			"Could not handle %s operation for %s because ServiceInstance %s is marked for deletion",
+			req.Operation,
+			req.Kind.Kind,
+			instanceRef.Name)
+		return fmt.Errorf("could not %s %s %q", req.Operation, sb.Kind, sb.Name)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handle_reference.go]: we have a contract for external errors which are returned for users, please change that and return the same error as it was originally. Additionally from your error msg user will not find out what was the problem

File: pkg/webhook/servicecatalog/servicebinding/validation/handle_reference_test.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+			err := handler.InjectDecoder(decoder)
+			require.NoError(t, err)
+			err = handler.InjectClient(fakeClient)
+			require.NoError(t, err)
+
+			request.AdmissionRequest.Operation = test.operation
+
+			// When
+			response := handler.Handle(context.TODO(), request)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handle_reference_test.go]:
```suggestion
			response := handler.Handle(context.Background(), request)
```
`context.TODO()`  should be used only if you know that in future you will need to replace it with some more production impl. Here you will not do that.

same suggestion for all other occurrences in tests cases

File: pkg/webhook/servicecatalog/servicebinding/validation/handler.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+// Validator is used to implement new validation logic
+type Validator interface {
+	Validate(context.Context, admission.Request, *sc.ServiceBinding, *webhookutil.TracedLogger) error
+}
+
+// AdmissionHandler handles ServiceBinding validation
+type AdmissionHandler struct {
+	decoder *admission.Decoder
+	client  client.Client
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handler.go]: not used, can be removed

same suggestion for all other places

File: pkg/webhook/servicecatalog/servicebroker/validation/handler_access.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	if !sar.Status.Allowed {
+		traced.Infof(
+			"Could not handle %s operation for %s %q because SubjectAccessReview has allowed status set to false",
+			req.Operation,
+			sb.Kind,
+			sb.Name,
+		)
+		return fmt.Errorf("could not %s %s %q", req.Operation, sb.Kind, sb.Name)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebroker/validation/handler_access.go]: please return the same error msg as it was originally. Your error does not say what was the problem.

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	if !sar.Status.Allowed {
+		traced.Infof(
+			"Could not handle %s operation for %s %q because SubjectAccessReview has allowed status set to false",
+			req.Operation,
+			csb.Kind,
+			csb.Name,
+		)
+		return fmt.Errorf("could not %s %s %q", req.Operation, csb.Kind, csb.Name)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go]: please return the same error msg as it was originally. Your error does not say what was the problem.

File: pkg/webhook/servicecatalog/servicebinding/validation/handler.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+var _ admission.Handler = &AdmissionHandler{}
+var _ admission.DecoderInjector = &AdmissionHandler{}
+var _ inject.Client = &AdmissionHandler{}
+
+// NewAdmissionHandler creates new AdmissionHandler and initializes validators list
+func NewAdmissionHandler() *AdmissionHandler {
+	return &AdmissionHandler{
+		CreateValidators: []Validator{&ReferenceDeletion{}},
+		UpdateValidators: []Validator{&ReferenceDeletion{}},
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handler.go]: it shouldn't be registered for update action. Remember that we are porting functionality and the behavior need to be the same as it was previously. Right now we do not want to add any new functionality or do any bug fixing.

**Review by mszostok — CHANGES_REQUESTED** [review]
FYI: Michał added `MutateError` on his PR https://github.com/kyma-incubator/service-catalog/pull/10/files#diff-18ff0721b44f24b4ea6b743c8f4ac6e2 

IMO this needs to be it must be generalized and moved to `webhookutil` and you should also use that in your validators and then in handler

Let's discuss that f2f

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go (hunk 1 above)
Comment (jasiu001) [reply pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]: Client is used to create `SubjectAccessReview` in `pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go`

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go (hunk 2 above)
Comment (jasiu001) [reply pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]: I tried to be consistent with this [webhook](https://github.com/kyma-incubator/service-catalog/pull/14/files#diff-5bec77c449efcecb597db56e3fd2a118R61). If we want to change one of the log then I suggest change all

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go (hunk 3 above)
Comment (jasiu001) [reply pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access.go]: We provide an injection of client in to validation struct. This [method](https://github.com/kyma-incubator/service-catalog/pull/14/files#diff-5bec77c449efcecb597db56e3fd2a118R117) handle it. One more time, if we want to add this we should be consistent with all validator webhooks.

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go (hunk 5 above)
Comment (jasiu001) [reply pkg/webhook/servicecatalog/clusterservicebroker/validation/handler_access_test.go]: it is 😄

File: pkg/webhook/servicecatalog/servicebinding/validation/handler.go (hunk 13 above)
Comment (jasiu001) [reply pkg/webhook/servicecatalog/servicebinding/validation/handler.go]: [Comment](https://github.com/kubernetes-incubator/service-catalog/blob/v0.1.41/plugin/pkg/admission/servicebindings/lifecycle/admission.go#L47) suggests it should be also check for update action

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+		admission.InjectDecoderInto(d, v)
+	}
+
+	return nil
+}
+
+// InjectClient injects the client into the handlers
+func (h *AdmissionHandler) InjectClient(c client.Client) error {
+	for _, v := range h.CreateValidators {
+		inject.ClientInto(c, v)
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]: you should check the error and return it. (return fast it's ok here) 

Same suggestion for other usecases

File: pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		traced.Infof("ClusterServiceBroker validation wehbook does not support action %q", req.Operation)
+		return admission.Allowed("action not taken")
+	}
+
+	if err != nil {
+		switch err.Code() {
+		case http.StatusForbidden:
+			return admission.Denied(err.Error())
+		default:
+			return admission.Errored(err.Code(), errors.New(err.Error()))
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/clusterservicebroker/validation/handler.go]:
```suggestion
			return admission.Errored(err.Code(), err)
```

File: pkg/webhook/servicecatalog/servicebinding/validation/handler.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		traced.Infof("ServiceBinding validation wehbook does not support action %q", req.Operation)
+		return admission.Allowed("action not taken")
+	}
+
+	if err != nil {
+		switch err.Code() {
+		case http.StatusForbidden:
+			return admission.Denied(err.Error())
+		default:
+			return admission.Errored(err.Code(), errors.New(err.Error()))
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebinding/validation/handler.go]:
```suggestion
			return admission.Errored(err.Code(), err)
```

File: pkg/webhook/servicecatalog/servicebroker/validation/handler.go [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+		traced.Infof("ServiceBroker validation wehbook does not support action %q", req.Operation)
+		return admission.Allowed("action not taken")
+	}
+
+	if err != nil {
+		switch err.Code() {
+		case http.StatusForbidden:
+			return admission.Denied(err.Error())
+		default:
+			return admission.Errored(err.Code(), errors.New(err.Error()))
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/servicebroker/validation/handler.go]:
```suggestion
			return admission.Errored(err.Code(), err)
```

File: pkg/webhook/servicecatalog/serviceinstance/validation/handler.go [hunk 18]
```diff
@@ -71,31 +70,40 @@ func (h *AdmissionHandler) Handle(ctx context.Context, req admission.Request) ad
...
 	}
 
-	if errs != nil {
-		return admission.Denied(errs.Error())
+	if err != nil {
+		switch err.Code() {
+		case http.StatusForbidden:
+			return admission.Denied(err.Error())
+		default:
+			return admission.Errored(err.Code(), errors.New(err.Error()))
```
Comment (mszostok) [inline pkg/webhook/servicecatalog/serviceinstance/validation/handler.go]:
```suggestion
			return admission.Errored(err.Code(), err)
```

**Review by mszostok — APPROVED** [review]

**Pull request merged**